firmware/test/application/test_file_reader.cpp

```diff
@@ -0,0 +1,155 @@
+    int line_count = 0;
+    for (const auto& line : reader) {
+        printf("Line: %s", line.c_str());
+        ++line_count;
+    }
+
+    CHECK_EQ(line_count, 3);
+}
+
+TEST_CASE("It can iterate file ending with newline.") {
+    MockFile f{"abc\ndef\nhij\n"};
+    BufferLineReader<MockFile> reader{f};
+    int line_count = 0;
+    for (const auto& line : reader) {
+        printf("Line: %s", line.c_str());
+        ++line_count;
+    }
+
+    CHECK_EQ(line_count, 3);
+}
+
+TEST_CASE("It can iterate files with empty lines.") {
+    MockFile f{"abc\ndef\n\nhij\n\n"};
+    BufferLineReader<MockFile> reader{f};
+    int line_count = 0;
+    for (const auto& line : reader) {
+        printf("Line: %s", line.c_str());
+        ++line_count;
+    }
+
+    CHECK_EQ(line_count, 5);
+}
+
+TEST_CASE("It can iterate large lines.") {
+    std::string long_line(0x90, 'b');
+    long_line.back() = 'c';
+
+    MockFile f{long_line + "\n" + long_line + "\n"};
+    BufferLineReader<MockFile> reader{f};
+    int line_count = 0;
+    for (const auto& line : reader) {
+        CHECK_EQ(line.length(), 0x91);
+        printf("Line: %s", line.c_str());
+        ++line_count;
+    }
+
+    CHECK_EQ(line_count, 2);
+}
+
+TEST_SUITE_END();
+
+TEST_SUITE_BEGIN("Test split_string");
+
+TEST_CASE("Empty string returns no results.") {
+    auto r = split_string("", ',');
+    CHECK(r.empty());
+}
+
+TEST_CASE("String without delimiter returns 1 result.") {
+    auto r = split_string("hello", ',');
+    REQUIRE_EQ(r.size(), 1);
+    CHECK(r[0] == "hello");
+}
+
+TEST_CASE("It will split on delimiter.") {
+    auto r = split_string("hello,world", ',');
+    REQUIRE_EQ(r.size(), 2);
+    CHECK(r[0] == "hello");
+    CHECK(r[1] == "world");
+}
+
+TEST_CASE("It will return empty columns.") {
+    auto r = split_string("hello,,world", ',');
+    REQUIRE_EQ(r.size(), 3);
+    CHECK(r[0] == "hello");
+    CHECK(r[1] == "");
+    CHECK(r[2] == "world");
+}
+
+TEST_CASE("It will return empty first column.") {
+    auto r = split_string(",hello,world", ',');
+    REQUIRE_EQ(r.size(), 3);
+    CHECK(r[0] == "");
+    CHECK(r[1] == "hello");
+    CHECK(r[2] == "world");
+}
+
+TEST_CASE("It will return empty last column.") {
+    auto r = split_string("hello,world,", ',');
+    REQUIRE_EQ(r.size(), 3);
+    CHECK(r[0] == "hello");
+    CHECK(r[1] == "world");
+    CHECK(r[2] == "");
+}
+
+TEST_CASE("It will split only empty columns.") {
+    auto r = split_string(",,,,", ',');
+    REQUIRE_EQ(r.size(), 5);
+    CHECK(r[0] == "");
+    CHECK(r[4] == "");
+}
+
+TEST_SUITE_END();
+
+/* Simple example of how to use this to read settings by lines. */
+TEST_CASE("It can parse a settings file.") {
+    MockFile f{"100,File.txt,5\n200,File2.txt,7"};
+    BufferLineReader<MockFile> reader{f};
+    std::vector<std::string> data;
+
+    for (const auto& line : reader) {
+        auto cols = split_string(line, ',');
+        for (auto col : cols)
+            data.emplace_back(col);
```

Comment:
In a real scenario, you'd do your actual parsing here - I'm just putting them in to a vector for the test.